aten/src/ATen/ParallelOpenMP.h

```diff
@@ -25,16 +25,18 @@ inline void parallel_for(
 #ifdef _OPENMP
   std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
   std::exception_ptr eptr;
-  // choose number of tasks based on grain size and number of threads
-  int64_t num_threads = omp_in_parallel() ? 1 : omp_get_max_threads();
-  if (grain_size > 0) {
-    num_threads = std::min(num_threads, divup((end - begin), grain_size));
-  }
 
-#pragma omp parallel num_threads(num_threads)
+#pragma omp parallel if (!omp_in_parallel() && ((end - begin) > grain_size))
   {
+    // choose number of tasks based on grain size and number of threads
+    // can't use num_threads clause due to bugs in GOMP's thread pool (See #32008)
+    int64_t num_threads = omp_get_num_threads();
+    if (grain_size > 0) {
+      num_threads = std::min(num_threads, divup((end - begin), grain_size));
+    }
+
     int64_t tid = omp_get_thread_num();
-    int64_t chunk_size = divup((end - begin), omp_get_num_threads());
```

Comment:
This line in the original code was incorrect.
Since the work is to be distributed among `num_threads` threads, This line should have been 
`int64_t chunk_size = divup((end - begin), num_threads);`

That's why @CaoZhongZ's [workaround patch](https://github.com/CaoZhongZ/pytorch/pull/1/files) delivered correct results.

Reply:
`omp_get_num_threads()` corresponds to the number of threads in the active parallel region. See these docs:
https://gcc.gnu.org/onlinedocs/libgomp/omp_005fget_005fnum_005fthreads.html

Specifically note the reference to `NUM_THREADS` clause:

>At runtime, the size of the current team may be set either by the NUM_THREADS clause or by omp_set_num_threads.

Reply:
Oh, thanks!